### PR TITLE
[1.20.4] Add Mob Split Event

### LINKED ISSUE #521: Add event for Slime splitting into smaller slimes
This event would act an an alternative for LivingConversionEvent but instead for any one-to-many actions. Such as Slime splitting into many new smaller Slime entities. Said event would allow modders to explicitly opt in and copy the exact data from parent to the many new entities. Such as if parent is paralyzed, mod may want to copy that to the new smaller slimes as well.

Would handle the edge case that this PR cannot cover: https://github.com/neoforged/NeoForge/pull/514

also consider Magma Cube splitting and Skeleton Trap spawning many horsemen (if this is even applicable)

### PR DESCRIPTION
**Pull request by Shadows-of-Fire**
This PR adds the `MobSplitEvent`, which is fired when a dying slime (or magma cube, or other modded thing) creates child mobs.  It provides access to the child and the parent to allow copying data, and can be cancelled to prevent the children from spawning.

Closes #521

### CODE REVIEW

**Review by TheCurle — DISMISSED** [review]
Nice and simple.

**Comment by TelepathicGrunt** [comment]
What about Magma Cubes?

**Comment by Shadows-of-Fire** [comment]
`public class MagmaCube extends Slime`
Already handled :p

**Comment by TelepathicGrunt** [comment]
Ah alright. The event name made it sound like it excluded Magma Cube. Not sure if there’s a better name to make it sound Ike it should be used for any entity that splits into more of themselves

**Comment by Shadows-of-Fire** [comment]
Updated so that only one event is fired with access to all child slimes.

**Comment by Gaming32** [comment]
> Ah alright. The event name made it sound like it excluded Magma Cube. Not sure if there’s a better name to make it sound Ike it should be used for any entity that splits into more of themselves

I personally think of Magma Cubes as Nether slimes. They even used to be named Magma Slimes.

File: src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+    /**
+     * Gets the parent slime that is in the process of being removed.
+     */
+    public Slime getParent() {
+        return parent;
+    }
+
+    /**
+     * Gets the mutable list of all child slimes.
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java]: You can use an inline return tag.

File: src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    protected boolean cancelled = false;
+
+    public SlimeSplitEvent(Slime parent, List<Slime> children) {
+        this.parent = parent;
+        this.children = children;
+    }
+
+    /**
+     * Gets the parent slime that is in the process of being removed.
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java]: You can use an inline return tag.

File: src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+ * This event is fired whenever a slime is removed and splits into multiple child slimes.
+ * <p>
+ * It is called from {@link Slime#remove(RemovalReason)}, and is only fired on the logical server.
+ */
+public class SlimeSplitEvent extends Event implements ICancellableEvent {
+
+    protected final Slime parent;
+    protected final List<Slime> children;
+
+    protected boolean cancelled = false;
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java]: This field is useless?

File: src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    public SlimeSplitEvent(Slime parent, List<Slime> children) {
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java]: Should be marked as `@Internal`

**Review by Matyrobbrt — CHANGES_REQUESTED** [review]

File: src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+package net.neoforged.neoforge.event.entity.living;
+
+import java.util.List;
+import net.minecraft.world.entity.Entity.RemovalReason;
+import net.minecraft.world.entity.monster.Slime;
+import net.neoforged.bus.api.Event;
+import net.neoforged.bus.api.ICancellableEvent;
+
+/**
+ * This event is fired whenever a slime is removed and splits into multiple child slimes.
```
Comment (Technici4n) [inline src/main/java/net/neoforged/neoforge/event/entity/living/SlimeSplitEvent.java]: I think it's worth adding that this fires for magma cubes and potentially other subclasses of Slime too.

**Comment by TelepathicGrunt** [comment]
Copying my last comments from Discord to here:

The current event requires all entities that fire this event to extend Slime class which may not be true for modded slime mods that using their own class that doesn't extend Slime but still has the splitting behavior.

As long as the event’s type is Mob or LivingEntity instead of Slime, it tells people that the event could be firing for modded non-slime derived class and to do a check on the entity. That’s why the event name would be ideally renamed to maybe LivingEntitySplittingEvent. 

After all, we don’t have the issue of consumers of LivingEntityTickEvent not expecting it to be firing for all kinds of entities because the event name and typing already tells consumer it is broad and to do their own filtering inside.

**Comment by Shadows-of-Fire** [comment]
Type bounds loosened to `Mob`.

**Review by TelepathicGrunt — APPROVED** [review]

File: src/main/java/net/neoforged/neoforge/event/entity/living/MobSplitEvent.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+    /**
+     * {@return the parent mob, which is in the process of being removed.}
+     */
+    public Mob getParent() {
+        return parent;
+    }
+
+    /**
+     * {@return the mutable list of all children.}
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/MobSplitEvent.java]: Inline returns have no full stop at the end.

File: src/main/java/net/neoforged/neoforge/event/entity/living/MobSplitEvent.java [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+     * @apiNote Use {@link EventHooks#onMobSplit(Slime, List)} to fire this event.
+     */
+    @ApiStatus.Internal
+    public MobSplitEvent(Mob parent, List<Mob> children) {
+        this.parent = parent;
+        this.children = children;
+    }
+
+    /**
+     * {@return the parent mob, which is in the process of being removed.}
```
Comment (Matyrobbrt) [inline src/main/java/net/neoforged/neoforge/event/entity/living/MobSplitEvent.java]: Inline returns have no full stop at the end.

**Review by Matyrobbrt — DISMISSED** [review]

**Review by shartte — APPROVED** [review]

**Review by embeddedt — APPROVED** [review]

**Pull request merged**